core/src/main/scala/org/bitcoins/core/psbt/PSBTHelper.scala

```diff
@@ -0,0 +1,256 @@
+import org.bitcoins.core.util.BitcoinScriptUtil
+import scodec.bits.ByteVector
+
+import scala.annotation.tailrec
+
+object PSBTHelper {
+
+  final private val terminator: Byte = 0x00.byteValue
+
+  private def parseKeyAndValue(
+      bytes: ByteVector): (Long, ByteVector, ByteVector) = {
+    val keyCmpctUInt = CompactSizeUInt.parse(bytes)
+
+    if (keyCmpctUInt.toLong == 0) {
+      (keyCmpctUInt.size, ByteVector.empty, ByteVector.empty)
+    } else {
+      val key = bytes.drop(keyCmpctUInt.size).take(keyCmpctUInt.toLong)
+      val valueBytes = bytes.drop(keyCmpctUInt.size + keyCmpctUInt.toLong)
+      val valueCmpctUInt = CompactSizeUInt.parse(valueBytes)
+      val value = valueBytes
+        .drop(valueCmpctUInt.size)
+        .take(valueCmpctUInt.toLong)
+      val totalSize = keyCmpctUInt.size + keyCmpctUInt.toLong + valueCmpctUInt.size + valueCmpctUInt
+        .toLong
+
+      (totalSize, key, value)
+    }
+  }
+
+  @tailrec
+  def parseGlobalMap(
+      remainingBytes: ByteVector,
+      accum: Seq[GlobalPSBTRecord],
+      txOpt: Option[Transaction]): (
+      ByteVector,
+      Vector[GlobalPSBTRecord],
+      Option[Transaction]) = {
+    if (remainingBytes.head == terminator) {
+      (remainingBytes.tail, accum.toVector, txOpt)
+    } else {
+
+      val (totalSize, key, value) = parseKeyAndValue(remainingBytes)
+      val next = remainingBytes.drop(totalSize) // Drop the key-value map
+
+      PSBTGlobalKey.fromByte(key.head) match {
+        case UnsignedTransactionKey =>
+          txOpt match {
+            case Some(_) =>
+              throw new Exception("Duplicate Key, unsigned tx already provided")
+            case None =>
+              require(
+                key.size == 1,
+                s"The key must only contain the 1 byte type, got: ${key.size}")
+
+              val transaction: Transaction = Transaction.fromBytes(value)
+              parseGlobalMap(next,
+                             accum :+ UnsignedTransaction(transaction),
+                             Option(transaction))
+          }
+        case XPubKeyKey =>
+          val xpub = ExtPublicKey.fromBytes(key.tail.take(78))
+          val fingerprint = value.take(4)
+          val path = BIP32Path.fromBytes(value.drop(4))
+          parseGlobalMap(next, accum :+ XPubKey(xpub, fingerprint, path), txOpt)
+        case VersionKey =>
+          require(
+            key.size == 1,
+            s"The key must only contain the 1 byte type, got: ${key.size}")
+
+          val version = UInt32(value)
+          parseGlobalMap(next, accum :+ Version(version), txOpt)
+        case _ =>
+          parseGlobalMap(next,
+                         accum :+ GlobalPSBTRecord.Unknown(key, value),
+                         txOpt)
+      }
+    }
+  }
+
+  @tailrec
+  def parseInputMaps(
+      remainingBytes: ByteVector,
+      remainingInputs: Int,
+      accum: Seq[InputPSBTMap]): (ByteVector, Vector[InputPSBTMap]) = {
```

Comment:
I think we should introduce a 
```private[psbt] case class PSBTParseResult[T](remainingBytes: ByteVector, parsed: T)```
and use that as the return type on all of the functions in this file rather than using tuples.
So this return type becomes `PSBTParseResult[Vector[InputPSBTMap]]` for example.
And the global one becomes `PSBTParseResult[(Vector[GlobalPSBTRecord], Option[Transaction])]` and so on